altair/datasets/_readers.py

```diff
@@ -0,0 +1,569 @@
+
+        Effectively an eager read, no filters.
+        """
+        return (
+            nw.from_native(self.scan_fn(_METADATA)(_METADATA))
+            .lazy()
+            .collect()
+            .to_native()
+        )
+
+    @property
+    def _cache(self) -> Path | None:  # type: ignore[return]
+        """
+        Returns path to datasets cache, if possible.
+
+        Requires opt-in via environment variable::
+
+            Reader._ENV_VAR
+        """
+        if _dir := os.environ.get(self._ENV_VAR):
+            cache_dir = Path(_dir)
+            cache_dir.mkdir(exist_ok=True)
+            return cache_dir
+
+    def _import(self, name: str, /) -> Any:
+        if spec := find_spec(name):
+            return import_module(spec.name)
+        else:
+            reqs = _requirements(self._name)  # type: ignore[call-overload]
+            if isinstance(reqs, tuple):
+                depends = ", ".join(f"{req!r}" for req in reqs) + " packages"
+            else:
+                depends = f"{reqs!r} package"
+
+            msg = (
+                f"Backend {self._name!r} requires the {depends}, but {name!r} could not be found.\n"
+                f"This can be installed with pip using:\n"
+                f"    pip install {name}\n"
+                f"Or with conda using:\n"
+                f"    conda install -c conda-forge {name}"
+            )
+            raise ModuleNotFoundError(msg, name=name)
+
+    def __repr__(self) -> str:
+        return f"Reader[{self._name}]"
+
+    def __init__(self, name: LiteralString, /) -> None: ...
+
+
+class _PandasReader(_Reader["pd.DataFrame", "pd.DataFrame"]):
+    def __init__(self, name: _Pandas, /) -> None:
+        self._name = _requirements(name)
+        if not TYPE_CHECKING:
+            pd = self._import(self._name)
+        self._read_fn = {
+            ".csv": pd.read_csv,
+            ".json": pd.read_json,
+            ".tsv": partial["pd.DataFrame"](pd.read_csv, sep="\t"),
+            ".arrow": pd.read_feather,
+            ".parquet": pd.read_parquet,
+        }
+        self._scan_fn = {".parquet": pd.read_parquet}
+
+
+class _PandasPyArrowReader(_Reader["pd.DataFrame", "pd.DataFrame"]):
+    def __init__(self, name: Literal["pandas[pyarrow]"], /) -> None:
+        _pd, _pa = _requirements(name)
+        self._name = name
+        if not TYPE_CHECKING:
+            pd = self._import(_pd)
+            pa = self._import(_pa)  # noqa: F841
+
+        self._read_fn = {
+            ".csv": partial["pd.DataFrame"](pd.read_csv, dtype_backend=_pa),
+            ".json": partial["pd.DataFrame"](pd.read_json, dtype_backend=_pa),
+            ".tsv": partial["pd.DataFrame"](pd.read_csv, sep="\t", dtype_backend=_pa),
+            ".arrow": partial(pd.read_feather, dtype_backend=_pa),
+            ".parquet": partial(pd.read_parquet, dtype_backend=_pa),
+        }
+        self._scan_fn = {".parquet": partial(pd.read_parquet, dtype_backend=_pa)}
```

Comment:
@mattijn https://github.com/vega/altair/pull/3631#issuecomment-2484826592
> Regarding `backend=pandas["pyarrow"]`, I found it a bit unclear, especially since this option isn’t mentioned in the [pandas optional dependencies installation guide](https://pandas.pydata.org/docs/getting_started/install.html#optional-dependencies). The [pandas user guide](https://pandas.pydata.org/docs/user_guide/pyarrow.html#i-o-reading) uses `engine="pyarrow"`, so opting for this for consistency.

Appreciate the suggestion for consistency, but I think this would be *overloading* the term `engine`.

For `pandas`, we're just using [the default](https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html) of `engine="auto"`.

In the `polars` case, this instead corresponds to `use_pyarrow=True`

<details><summary><code>"polars[pyarrow]"</code></summary>
<p>

https://github.com/vega/altair/blob/7b3a89e5b5374eb391b7ae73ace219327069f979/altair/datasets/_readers.py#L339-L388

</p>
</details> 

---

### Read functions



| extension | `polars`                                                                                         | `pandas`                                                                                 | `pyarrow `                                                                                                  |
| --------- | ------------------------------------------------------------------------------------------------ | ---------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------- |
| .csv      | [pl.read_csv](https://docs.pola.rs/api/python/stable/reference/api/polars.read_csv.html)         | [pd.read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)         | [pa.csv.read_csv](https://arrow.apache.org/docs/python/generated/pyarrow.csv.read_csv.html)                 |
| .json     | [pl.read_json](https://docs.pola.rs/api/python/stable/reference/api/polars.read_json.html)       | [pd.read_json](https://pandas.pydata.org/docs/reference/api/pandas.read_json.html)       | [pa.json.read_json](https://arrow.apache.org/docs/python/generated/pyarrow.json.read_json.html)             |
| .arrow    | [pl.read_ipc](https://docs.pola.rs/api/python/stable/reference/api/polars.read_ipc.html)         | [pd.read_feather](https://pandas.pydata.org/docs/reference/api/pandas.read_feather.html) | [pa.feather.read_feather](https://arrow.apache.org/docs/python/generated/pyarrow.feather.read_feather.html) |
| .parquet  | [pl.read_parquet](https://docs.pola.rs/api/python/stable/reference/api/polars.read_parquet.html) | [pd.read_parquet](https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html) | [pa.parquet.read_table](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html)     |

#### Summary
- `polars`, `pyarrow`
	- None accept an `engine` argument
- `pandas`
	- 3/4 accept an `engine` argument
	- *What* that `engine` can be is different for each function

---

@mattijn the way I'd conceptualise this is in terms of **optional-dependencies**:

https://github.com/pandas-dev/pandas/blob/6a7685faf104f8582e0e75f1fae58e09ae97e2fe/pyproject.toml#L60-L62

https://github.com/pola-rs/polars/blob/f7c6bcc0f2b27e77ea70f429ad6d878f4cf7e724/py-polars/pyproject.toml#L39-L44

Both `"pandas[pyarrow]"` and `"polars[pyarrow]"` are **already** ways that you can install these packages with extras.

I'm using `packaging.requirements.Requirement`, with some additional checks to validate this here:

https://github.com/vega/altair/blob/7b3a89e5b5374eb391b7ae73ace219327069f979/altair/datasets/_readers.py#L540-L569

I also gave some examples in:
**1. How does this translate beyond single package backends?** from https://github.com/vega/altair/pull/3631#discussion_r1847091550

We *could* (although I'm not proposing we do this now) then add these as dependency groups to make installation even easier.
I think this would have more value when expanding to the spatial packages.

Reply:
For me, these are backend-specific details. If `engine` is something pandas users already know, and we want to support it, we can make options like `auto` or `pyarrow` available in `.to_pandas()`. If `use_pyarrow` is common for polars users, and we want that option, we can add it in `.to_polars()`. I don’t want to invent a new standard, I prefer sticking to what users already understand.